### PR TITLE
feat(ai): ai-proxy add more apis

### PR DESCRIPTION
**Pull request by sfwn**
#### What this PR does / why we need it:

ai-proxy add more apis

#### Which issue(s) this PR fixes:

- [Erda Cloud Issue Link](https://erda.cloud/terminus/dop/projects/1265/issues/all?id=495544&iterationID=12492&tab=REQUIREMENT&type=REQUIREMENT)

#### Specified Reviewers:

/assign @wangzhuzhen 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |    ai-proxy add more apis          |
| 🇨🇳 中文    |   ai-proxy 提供更多 apis            |

#### Need cherry-pick to release versions?

No.

### CODE REVIEW

File: .erda/ai-proxy/migrations/ai-proxy/20230823-ai-proxy.sql [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    `id`         CHAR(36) NOT NULL COMMENT 'primary key',
+    `created_at` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT '创建时间',
+    `updated_at` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP COMMENT '更新时间',
+    `deleted_at` DATETIME NOT NULL DEFAULT '1970-01-01 00:00:00' COMMENT '删除时间, 1970-01-01 00:00:00 表示未删除',
+
+    `client_id`  CHAR(36) NOT NULL COMMENT '客户端 id',
+    `model_id`   CHAR(36) NOT NULL COMMENT '模型 id',
+
+    PRIMARY KEY (`id`),
+    UNIQUE INDEX `unique_clientid_modelid` (`client_id`, `model_id`, `deleted_at`) # 一个模型在一个客户端下只能被关联一次。即使实际在物理上是同一个 model，但是在 model 表里也会有多条记录（model_type 不同）。
```
Comment (wangzhuzhen) [inline .erda/ai-proxy/migrations/ai-proxy/20230823-ai-proxy.sql:70]: which table has filed model_type?

File: internal/apps/ai-proxy/filters/azure-director/filter.go [hunk 2]
```diff
@@ -127,42 +127,42 @@ func (f *AzureDirector) FindProcessor(ctx context.Context, processor string) (fu
...
 	if !ok || value == nil {
 		return errors.New("provider not set in context map")
 	}
-	prov := value.(*models.AIProxyProviders)
+	prov := value.(*model_provider.ModelProvider)
 	reverseproxy.AppendDirectors(ctx, func(req *http.Request) {
-		if appKey := prov.GetAPIKeyWithDecrypt(); appKey != "" && req.Header.Get("Authorization") == "" {
-			req.Header.Set("Authorization", vars.ConcatBearer(appKey))
+		if apiKey := prov.APIKey; apiKey != "" && req.Header.Get("Authorization") == "" {
+			req.Header.Set("Authorization", vars.ConcatBearer(apiKey))
```
Comment (wangzhuzhen) [inline internal/apps/ai-proxy/filters/azure-director/filter.go:137]: Is apiKey need encrypted in header Authorization？

File: internal/apps/ai-proxy/filters/azure-director/filter.go [hunk 3]
```diff
@@ -127,42 +127,42 @@ func (f *AzureDirector) FindProcessor(ctx context.Context, processor string) (fu
...
 	if !ok || value == nil {
 		return errors.New("provider not set in context map")
 	}
-	prov := value.(*models.AIProxyProviders)
+	prov := value.(*model_provider.ModelProvider)
 	reverseproxy.AppendDirectors(ctx, func(req *http.Request) {
-		if appKey := prov.GetAPIKeyWithDecrypt(); appKey != "" && req.Header.Get("Api-Key") == "" {
-			req.Header.Set("Api-Key", appKey)
+		if apiKey := prov.APIKey; apiKey != "" && req.Header.Get("Api-Key") == "" {
+			req.Header.Set("Api-Key", apiKey)
```
Comment (wangzhuzhen) [inline internal/apps/ai-proxy/filters/azure-director/filter.go:165]: Is apiKey need encrypted in header Api-Key？

File: internal/apps/ai-proxy/models/metadata/metadata.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	}
+	return string(b), nil
+}
+
+func (m *Metadata) MergeMap() map[string]string {
+	if m == nil {
+		return nil
+	}
+	result := make(map[string]string)
+	for k, v := range m.Public {
```
Comment (wangzhuzhen) [inline internal/apps/ai-proxy/models/metadata/metadata.go:59]: Overwrite may be  happened in result?

File: internal/apps/ai-proxy/models/prompt/dbclient.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+		ClientID: req.ClientId,
+		Messages: message.FromProtobuf(req.Messages),
+		Metadata: metadata.FromProtobuf(req.Metadata),
+	}
+	if err := tx.Model(c).Create(c).Error; err != nil {
+		tx.Rollback()
+		return nil, err
+	}
+	// commit
+	tx.Commit()
```
Comment (wangzhuzhen) [inline internal/apps/ai-proxy/models/prompt/dbclient.go:56]: Is tx.Commit() need to be call before exit when some error happens?

**Review by wangzhuzhen — COMMENTED** [review]
big changes.

File: .erda/ai-proxy/migrations/ai-proxy/20230823-ai-proxy.sql (hunk 1 above)
Comment (sfwn) [reply .erda/ai-proxy/migrations/ai-proxy/20230823-ai-proxy.sql:71]: `ai_proxy_model`

File: internal/apps/ai-proxy/filters/azure-director/filter.go (hunk 2 above)
Comment (sfwn) [reply internal/apps/ai-proxy/filters/azure-director/filter.go:140]: No, the same as OpenAI and Azure OpenAI Service.

File: internal/apps/ai-proxy/filters/azure-director/filter.go (hunk 3 above)
Comment (sfwn) [reply internal/apps/ai-proxy/filters/azure-director/filter.go:168]: no.

File: internal/apps/ai-proxy/models/metadata/metadata.go (hunk 4 above)
Comment (sfwn) [reply internal/apps/ai-proxy/models/metadata/metadata.go:59]: yes, secret has higher priority.

File: internal/apps/ai-proxy/models/prompt/dbclient.go (hunk 5 above)
Comment (sfwn) [reply internal/apps/ai-proxy/models/prompt/dbclient.go:56]: No, `rollback` invoked if normal error occurred.

**Pull request merged**